pkg/gameserverallocations/controller.go

```diff
@@ -117,6 +122,32 @@ func NewController(wh *webhooks.WebHook,
 		},
 	})
 
+	agonesInformer.GameServers().Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
```

Comment:
I'm a little confused by this - I'm not seeing how this is better than the Informer cache. We still have to wait for the event to come through, so it is still eventually consistent -- and unless I'm mistaken (which is possible), this is a mirror of the Informer cache, so we don't gain anything?

Would it be easier / faster as a solution to have a list of potential Allocatable gameservers, and attempt to each on as allocated, and stop and return when one is successful? (fail on n number of attempts).

WDYT?

Reply:
@jkowalski WDYT on the above?

Reply:
cache.WaitForCacheSync(c.stop, c.gameServerSynced)  call in allocate() method seem to slow down the allocation call. Replacing with local cache and keep updating seem to improve the performance. The mutex blocking is only on the local cache instead of WaitForCacheSync.
The performance difference is not that much is the number of Ready servers is not changing (cluster is in steady state) and there are not too many concurrent allocation calls.

We don't need to wait on the event to come (at least for the Allocated Gameservers) since we are deleting the allocated servers from the cache.

I was thinking of the solution you suggested (in the second paragraph). I even try to implement something and tested but there are few moving parts where GSA , GS, FLT can easily go out of sync.

I am inspired from the Jarek's GameSerSet allocation and use of local cache.

Reply:
> cache.WaitForCacheSync(c.stop, c.gameServerSynced) call in allocate() method seem to slow down the allocation call.

Yep, makes perfect sense :+1: 

>  We don't need to wait on the event to come (at least for the Allocated Gameservers) since we are deleting the allocated servers from the cache.

But we aren't deleted gameservers from the cache when Fleets are scaled down - so we could  have race conditions as Fleets are scaled down.

> I was thinking of the solution you suggested (in the second paragraph). I even try to implement something and tested but there are few moving parts where GSA , GS, FLT can easily go out of sync.

Not sure what you mean here. Can you maybe explain it in a different way?

Reply:
> But we aren't deleted gameservers from the cache when Fleets are scaled down - so we could have race conditions as Fleets are scaled down.

This is true. However when/if the system tries to allocate a GS from cache that is deleted, it will fail to change the GS's state to Allocated since the revisonversion of the GS in the cache will be old.  

> Not sure what you mean here. Can you maybe explain it in a different way?

Since the GS state update ( from Ready to Allocated) and GSA record creation is not transnational ,we can end up creating multiple GSA records for the same GS.

Reply:
> This is true. However when/if the system tries to allocate a GS from cache that is deleted, it will fail to change the GS's state to Allocated since the revisonversion of the GS in the cache will be old.

Ah! true! But now we would likely return a 500 from the service when Fleets are scaling down. I think we need some retry logic.

> Since the GS state update ( from Ready to Allocated) and GSA record creation is not transnational ,we can end up creating multiple GSA records for the same GS.

I don't follow. My vision of the process would be:
1. Get n list of potential allocatable GS's (where n is essentially the set of retries)
1. Loop through list of GS's
   1. attempt to move to Allocated
   1. if successful, break
1. Update GSA with allocated GS details

Does that make sense?

Reply:
> I don't follow. My vision of the process would be:

1. Get n list of potential allocatable GS's (where n is essentially the set of retries)
2. Loop through list of GS's
   
   1. attempt to move to Allocated
   2. if not successful, continue to next GS
   3. if successful, break
3. Update GSA with allocated GS details

Does that make sense?

Makes sense. I would like to reproduce the condition and then try to implement what you are suggesting.